include/grpcpp/impl/codegen/server_callback.h

```diff
@@ -588,6 +622,14 @@ class CallbackClientStreamingHandler : public MethodHandler {
 
     reader->BindReactor(reactor);
     reactor->OnStarted(param.server_context, reader->response());
+    // Use release memory ordering on the cancellation state to make sure that
+    // all stores that took place in OnStarted are seen when the cancellation
+    // state is seen as being ON_STARTED_DONE. Not needed for failure case.
```

Comment:
Here and below, in the case of a failure, we call `ractor->OnCancel()`. The `OnCancel` implementation may need all of the changes made to reactor before this `cmpxchng`. So, I believe, we might acquire here. Is there a contract that `OnCancel()` shouldn't assume ordering?

Reply:
That's true. OnStarted may have triggered a StartRead which caused an OnReadDone etc, and there is nothing in the contract for OnCancel that limits its use. So we'll promote the failure case here (and elsewhere in this file) to memory_order_acquire

Reply:
I think my last comment wasn't correct since the acquire wouldn't control that anyway.

Reply:
So in the failure case here we're just executing code on the same thread as the one that just called OnStarted, so why would there be a reason to have acquire ordering? If operations triggered elsewhere (like OnReadDone as a result of StartRead being called form OnStarted), an acquire on this word wouldn't enforce a happens-before ordering there anyway, would it?

Reply:
Thank you. Oh, I see. Yes, changes in `OnStarted` are safe. My (rather stupid) question was whether `reactor` can be modified in a separate thread in parallel to this thread? In other words, can `ServerContext::CompletionOp::FinalizeResult` race with this thread?

I presumed the answer is yes, because otherwise you wouldn't have used atomics. Now, if that can happen, do we need the changes from the thread running `ServerContext::CompletionOp::FinalizeResult` be visible to this thread?

Reply:
Ah, thanks for the followup. Not a stupid question at all. So, I've noted in the comment on the other function: that function only produces 1 useful result into the Context, and that is the cancelled_ variable. However, that can only ever checked from within a lock (in the IsCancelled function), so I think that this atomic doesn't need further protection. Does that make sense?

Reply:
And the reactor can definitely be modified in another thread with OnReadDone, OnWritesDone, etc., but there is no guarantee of ordering for those with respect to each other or with respect to OnCancel.

Reply:
Oh I see. Thanks for the details. Yes, `relaxed` should work here, and `acquire` is needed on the other side. So, the code is correct as-is, AFAICT. Thanks

Reply:
Thanks for looking at this and providing comments!